dali/kernels/common/find/find_first_last_gpu_test.cu

```diff
@@ -0,0 +1,212 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/common/find/find_first_last_gpu.cuh"
+#include <gtest/gtest.h>
+#include <vector>
+#include "dali/core/cuda_event.h"
+#include "dali/kernels/dynamic_scratchpad.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/test/tensor_test_utils.h"
+#include "dali/test/test_tensors.h"
+
+namespace dali {
+namespace kernels {
+namespace find {
+namespace test {
+
+template <typename T>
+struct threshold {
+  T value_;
+  explicit threshold(T value = 0) : value_(value) {}
+
+  DALI_HOST_DEV DALI_FORCEINLINE bool operator()(T x) const noexcept {
+    return x >= value_;
+  }
+};
+
+template <typename T>
+void SequentialFillBothEnds(T* data, int64_t data_len, T start_value) {
+  auto value = start_value;
+  int i = 0;
+  int j = data_len - 1;
+  for (; i < j; value++) {
+    data[i++] = value;
+    data[j--] = value;
+  }
+  if (i == j)
+    data[i] = value;
+}
+
+class FindFirstLastTestGPU : public ::testing::Test {
+ public:
+  using T = float;
+  TestTensorList<T> in_;
+  TestTensorList<int64_t> out_begin_;
+  TestTensorList<int64_t> out_length_;
+
+  TestTensorList<int64_t> ref_begin_;
+  TestTensorList<int64_t> ref_length_;
+
+  using Predicate = threshold<T>;
+  Predicate thresh{3};
+  span<Predicate> predicates{&thresh, 1};
+
+  void SetUp() final {
+    int nsamples = 5;
+
+    // 1500 chosen so that it doesn't fit one CUDA block (32*32)
+    TensorListShape<> sh = {{5, }, {10, }, {9, }, {7, }, {1500, }};
+    TensorListShape<0> out_sh(nsamples);
+    in_.reshape(sh);
+    out_begin_.reshape(out_sh);
+    out_length_.reshape(out_sh);
+    ref_begin_.reshape(out_sh);
+    ref_length_.reshape(out_sh);
+
+    for (int i = 0; i < nsamples; i++) {
+      auto v = in_.cpu()[i];
+      SequentialFillBothEnds(v.data, static_cast<int64_t>(v.shape.num_elements()), T(0));
+    }
+
+    // Threshold = 3
+    // Input 0:  0 1 2 1 0
+    //           ^^
+    // Input 1:  0 1 2 3 4 4 3 2 1 0
+    //                 ^     ^
+    // Input 2:  0 1 2 3 4 3 2 1 0
+    //                 ^   ^
+    // Input 3:  0 1 2 3 2 1 0
+    //                 ^^
+    // Input 4:  0 1 2 3 4 5 ... 1494 1495 1496 1497 1498 1499
+    //                 ^                   ^
+    ref_begin_.cpu()[0].data[0] = 0;
+    ref_length_.cpu()[0].data[0] = 0;
+
+    ref_begin_.cpu()[1].data[0] = 3;
+    ref_length_.cpu()[1].data[0] = 4;
+
+    ref_begin_.cpu()[2].data[0] = 3;
+    ref_length_.cpu()[2].data[0] = 3;
+
+    ref_begin_.cpu()[3].data[0] = 3;
+    ref_length_.cpu()[3].data[0] = 1;
+
+    ref_begin_.cpu()[4].data[0] = 3;
+    ref_length_.cpu()[4].data[0] = 1494;
```

Comment:
Why it is not 1499, the last element?

Reply:
Because it's not the last element but the length of the region. In any case, the end of the region would be 1497 (note that the input is like a triangle:
```
0, 1, 2, 3, 4, ...., 4, 3, 2, 1, 0
         ^              ^
```

Reply:
Understood.